django/contrib/admin/options.py

```diff
@@ -1944,7 +1944,24 @@ def _create_formsets(self, request, obj, change):
                     'files': request.FILES,
                     'save_as_new': '_saveasnew' in request.POST
                 })
-            formsets.append(FormSet(**formset_params))
+            formset = FormSet(**formset_params)
+
+            def _should_delete(request, obj, formset, index):
+                """Return whether or not the user deleted the form."""
+                return (
+                    inline.has_delete_permission(request, obj) and
```

Comment:
Should audit tests carefully. For example, I don't see any failures when commenting `inline.has_delete_permission(request, obj)` out.

Reply:
I added an extra case to the `view_and_delete` case before adding the delete permission. This requires the `has_delete_permission()` check here. Thanks for the review!